lib/pengine/unpack.c

```diff
@@ -534,6 +534,24 @@ handle_startup_fencing(pe_working_set_t *data_set, pe_node_t *new_node)
     new_node->details->unseen = TRUE;
 }
 
+/*!
+ * \internal
+ * \brief Check whether a string is a valid XML ID
+ *
+ * \param[in] id  String to check
+ *
+ * \return true if \p id is a valid XML ID, or false otherwise
+ *
+ * \note See https://www.w3.org/TR/xml-id/#processing for the specification.
+ * \note Many valid Unicode characters are left unchecked. See \p NameStartChar
+ *       and \p NameChar here: https://www.w3.org/TR/REC-xml/#NT-Name.
```

Comment:
This is undesirable for non-English-speaking users. If they have Unicode characters in their node names, they'll be incorrectly warned. From what I can tell from cursory research, POSIX regexes have limited support for Unicode; character classes (e.g., `[:alpha:]`) will match Unicode characters if the locale is set properly. I don't know that we can explicitly specify as valid all the characters from the XML `NameStartChar` and `NameChar` spec. PCRE can probably do it but that'd be a new dependency.

In any case I haven't done any testing at all with Unicode yet.

Probably the ideal solution would be to make RelaxNG validate the ID for us. I can't find any function in `libxml2` or elsewhere that explicitly validates XML IDs, but we could create a minimal RelaxNG schema and a minimal XML node containing the ID, and pass it to a RelaxNG validator. We have code in `schemas.c` that validates XML in this basic way, but it's hard-coded for `pacemaker-X` schemas. So that approach would require either some larger changes in `schemas.c`, or duplicating here some of the code that sets us up to do the work.

Reply:
`libxml2` does provide `xmlValidateAttributeValue()`. That function validates IDs against the `Name` production. But that's incorrect. It should validate against the `NCName` production, which is a name with no colon character.

We could easily work around that by checking for a colon ourselves.
```
static inline bool
valid_xml_id(const char *id)
{
    return xmlValidateAttributeValue(XML_ATTRIBUTE_ID, (const xmlChar *) id)
           && (strchr(id, ':') == NULL);
}
```

We may also want to check for duplicate IDs and log a deprecation warning for that too. I think `libxml2:valid.c` provides functions that can take care of that for us. It is a bit more involved in terms of arguments, so I'd have to look more closely at what's required.

Reply:
That makes sense. Don't bother with the duplicate IDs. For 2.2.0, the plan is to have an XSL transform, which should be able to shuffle any non-conforming id's to cluster-id and auto-generate a unique ID.